### DISCUSSION TITLE
How to report the download progress of the dependencies?

### DISCUSSION THREAD

**Discussion by szabolcsdombi** [pyodide/pyodide#3772] [Q&A] [answered]
Hi, is there a way to report back to the user how the download is going?

```js
    await pyodide.loadPackage([
      'numpy',
      'scipy',
      'pillow',
      'zengl-1.12.1-cp311-cp311-emscripten_3_1_32_wasm32.whl',
      'webgl-0.1.0-cp311-cp311-emscripten_3_1_32_wasm32.whl',
      'custom1-1.0.0-cp311-cp311-emscripten_3_1_32_wasm32.whl',
      'custom2-0.1.0-py3-none-any.whl',
      'custom3-0.1.0-py3-none-any.whl',
      'custom4-0.1.0-py3-none-any.whl',
      'custom5-0.1.0-py3-none-any.whl',
      'custom6-0.1.0-py3-none-any.whl',
    ]);
 ```
    
 At the moment our loadPackage call downloads 200MB and it takes tens of seconds even with a good internet connection.
 I was wondering if reporting the download progress is implemented in some way.
 
An alternative would be to download these files manually with fetch, but then how can I load them?

**Comment by hoodmane** [comment]
Yeah it would be great to add a download progress API for `loadPackage`.

**Comment by ryanking13** [comment, marked as answer]
loadPackage consists of two steps: downloading packages + compiling WASM modules.

For downloading packages, we rely on `fetch`, so you can patch fetch to show progress (https://github.com/pyodide/pyodide/discussions/2927).

For compiling WASM modules, there isn't a clear way to show the progress AFAIK.

**Reply by szabolcsdombi** [reply]
this fully solves my problem, thank you.

**Comment by szabolcsdombi** [comment]
I changed the solution you linked to use an estimated total length and to support multiple downloads.
I paste it here because the original code was not working well when the `Content-Length` is not available (cors or compressed).

```js
const downloadedBytesEstimate = 17973279;
let downloadedBytes = 0;

const fetchOriginal = fetch;
const fetchWithProgress = async (url) => {
  const response = await fetchOriginal(url);

  const ts = new TransformStream({
    transform(chunk, ctrl) {
      downloadedBytes += chunk.byteLength;
      console.log('progress', Math.min(downloadedBytes / downloadedBytesEstimate, 1.0));
      ctrl.enqueue(chunk);
    }
  });

  return new Response(response.body.pipeThrough(ts), response);
};

window.fetch = fetchWithProgress;

const pyodide = await loadPyodide();

await pyodide.loadPackage([
  'zengl-1.12.2-cp311-cp311-emscripten_3_1_32_wasm32.whl',
  'zengl_webgl-0.1.0-cp311-cp311-emscripten_3_1_32_wasm32.whl',
  'canvas-0.1.0-cp311-cp311-emscripten_3_1_32_wasm32.whl',
  'vmath-0.1.0-cp311-cp311-emscripten_3_1_32_wasm32.whl',
  'game-0.1.0-cp311-cp311-emscripten_3_1_32_wasm32.whl',
  'assets-0.1.0-py3-none-any.whl',
]);

// console.log('downloadedBytesEstimate', downloadedBytes);
window.fetch = fetchOriginal;
```
